### ISSUE TITLE
Optimize multiple shared locks for a single process

### ISSUE THREAD

**Issue by icewind1991** [owncloud/core#19745]
When a single process acquires a multiple shared lock on the same file we only need to mark the file as locked in the database for the first lock, any additional shared locks only has to be kept track of within the process.

This would prevent unneeded db writes when doing locking

cc @PVince81 @rullzer

**Comment by rullzer** [comment]
So thinking  more about this makes me think about what the actual behaviour is we want.

Assume the following structure:

```
-dir
|-dir1
 |-dir1-1
  |-dir1-1-1
  |-dir1-1-2
 |-dir1-2
 |-dir1-3
```

Now I want to traverse `dir1` and all its children.

As I understand it currently. We ask for the content of `dir1`. To do this we lock `dir1` and `dir`. We then get the contents and unlock everything. Now I have an array with `dir1-1, dir1-2, dir1-3`. If I then want to get the contents of `dir1-1`, now to get that we lock `dir1-1`, `dir1` and `dir`. However there is no quarantee that there hasn't been a change.

So apart from saving a whole bunch of locking and unlocking operations... I think it would also be more correct to leave the read locks in place.

**Comment by icewind1991** [comment]
Fixed with #19890
